add structured Path objects, update joinpath: normalize, support url/net/windows

benfleis opened the pull request:
add structured Path objects, update joinpath: normalize, support url/net/windows

- Adds a Path class to provide structured support for ... paths! This includes local (posix/win, file:*) and various remote (URI, UNC). Provides non-optional normalization (. and .. removal where available/appropriate), but _not_ Canonicalization, and various helpers
- Improves FileSystem::JoinPath to normalize separators and dot segments, respect absolute RHS, preserve URI schemes, and handle windows drive prefixes.
- Other integrations are "minimal", in the sense that only conflicting/test-breaking code was updated to use this; there are many opportunities to expand its use in future PRs.
- Adds `path_join(arg1 [, ...])` function to SQL
- Testing is extensive in cpp land, compact in SQL land (basics + SQL specifics, especially NULL and empty string handling)

This is started by #19923, from which most of the SQL path_join code originates; thanks @feichai0017 !

pdet commented on test/sql/function/string/path_join.test:
```diff
@@ -0,0 +1,97 @@
+# name: test/sql/function/string/path_join.test
```
Since we are dealing with strings, it might be good to also test with very large strings due to the german-string representation (I think you are using it correctly btw, but always good to test :) )

pdet reviewed (commented):
Hey @benfleis Thanks for the PR, this is looking pretty awesome! 

I've added a few comments!

feichai0017 commented:
@benfleis Will you make any modifications to this branch?

benfleis commented:
I have an update in the wings, but was on the back burner due to our 1.4.4 and 1.5 releases; I actually already planned to pick it up this afternoon!

benfleis committed:
Squashed commit of the following:

commit e5d4d5ae80351a3549acc3de0eb27be8378a9f5e
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Mon Dec 22 18:22:00 2025 +0800

    Fix Windows glob path joining for drive roots

commit d6901efa3ce2ef3e41c20e45ba2493a9b01d2ffc
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Mon Dec 22 15:59:55 2025 +0800

    Adjust path_join null handling and tests

commit 4559885b9d414ed18fe07b2b93891493aacc1ce6
Merge: e1deafe411 c46a01b579
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Mon Dec 22 13:13:34 2025 +0800

    Merge branch 'main' into sanitize-joinpath

commit e1deafe4114746a2aa851e2e817fcb92f3e6846c
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:58 2025 +0800

    update function generation

commit 3dd5db31724a333aef2aa40856b0b5427ccabd27
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:57 2025 +0800

    correct code format

commit 2650670108dac0cbbd4893cf06283001624a5cd8
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:56 2025 +0800

    Fix JoinPath root separator duplication

commit f5e4d8a5b43ababa62813d741751a0e89a50dd29
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:55 2025 +0800

    Format path_join description

commit cbae45db47c3150e9e06b4b17a5c9eeaf695c9a1
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:54 2025 +0800

    Clarify path_join strictness and Windows path handling

commit 13a46b70147a4286943433a9644c58f57d7518c8
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:53 2025 +0800

    Fix file scheme joins with absolute paths

commit 15ffeb5c9071b9d767586fa9af7b17c4a04a1ed3
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:52 2025 +0800

    Fix JoinPath URI separator handling

commit eba9f05a30dba34626e25c252088671c936917d8
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:51 2025 +0800

    Force forward slashes for URI joins regardless of platform separator

commit b52fad951a4be7cdaa3c12882ed3e5004865bb7e
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:50 2025 +0800

    Normalize URI paths to forward slashes when separator is slash

commit a698a8a77809a9ea0e066a8a4a4760b5839419e3
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:49 2025 +0800

    Normalize URI paths to forward slashes in JoinPath

commit fd000090d81356928b4224ee7a24bfb9ed3c5eb7
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:48 2025 +0800

    Revert "Keep URI paths using forward slashes when joining"

    This reverts commit 9b06a6250e3548d09abf4bb8bac8dcbe825d0434.

commit 3bdc7df1799c995454a999c0ddd5b200e72f38e7
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:46 2025 +0800

    Keep URI paths using forward slashes when joining

commit 7c121626ae66ea2116248e7df2a863fcc692cbe1
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:15:26 2025 +0800

    Relax drive-relative clamp and tidy separator stripping

commit a45efab3b507b59af626a408d9904083b2fddfd9
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:12:13 2025 +0800

    Fix JoinPath file scheme handling

commit 2296b6c771d04a47165191ca5bc895881607270b
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sun Nov 30 02:39:51 2025 +1100

    Stabilize path_join evaluation

commit 832ccb5be8fcdbde538526e87d8f15737f1efe38
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sat Nov 29 23:33:29 2025 +1100

    Make path_join SQL test platform-neutral

commit 1666b8100ba9206c2dcb9efb418455e8e73503c6
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sat Nov 29 20:45:25 2025 +1100

    Regenerate function metadata for path_join

commit 4394f8e63661d62652536211ae1a978f8bd39990
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sat Nov 29 20:38:04 2025 +1100

    fix code format

commit bb4944a033d8a02549fe5aef6f8922486f6e540b
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sat Nov 29 20:36:00 2025 +1100

    Add path_join SQL function

commit dae576807248984e5b25ce4474f66297fb7f4dcd
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sat Nov 29 10:29:37 2025 +1100

    correct coede format

commit 41175dcd3e6069b6304b67c438453a947058bdca
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sat Nov 29 03:16:37 2025 +1100

    Fix Windows drive root joins and add glob regression test

commit 24cf2412f6f6daf1eae01684ea75c7935afd169c
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Fri Nov 28 22:42:49 2025 +1100

    Harden JoinPath normalization and edge tests

commit 909979054d0f14dc3a786cc4b2f5e85edd6e3989
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Tue Nov 25 17:09:50 2025 +1100

    fix code format

commit 2ae4636e734ad4be54b61833865daa830c321d58
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Tue Nov 25 16:53:21 2025 +1100

    Sanitize JoinPath handling and add tests

benfleis committed:
Squashed commit of PathJoin changes, plus fixups

Fixups include 2 tweaks to recent CanonicalizePath calls and
`allowed_directories` test, which relied on previous string-cat-like
JoinPath behavior. Ideally these functions should be fully integrated
with the parsing Path, but keep minimal for now.

commit 841ea42ea87cae3ebf5206773164fd4bf371afa6
Author: ben fleis <benfleis@users.noreply.github.com>
Date:   Sat Feb 21 19:08:25 2026 +0100

    cleanup: keep segments, stop extra copies

commit c4578ae8cd5955614d32127e6ca0510403ecd2fa
Author: ben fleis <benfleis@users.noreply.github.com>
Date:   Sat Feb 21 14:50:18 2026 +0100

    cleanups, pre segment keeping

commit 8c8f51829c8d72df9ee1e719c38d8cadb1b17391
Author: ben fleis <benfleis@users.noreply.github.com>
Date:   Sat Feb 21 12:37:04 2026 +0100

    fix integration & tests; still rename + some cleanups to do

commit e359d9ee31d5c7e8d905a6f0e23909e83621e804
Author: ben fleis <benfleis@users.noreply.github.com>
Date:   Tue Jan 6 08:45:24 2026 +0100

    exposed parse, need to cut back and finish

commit d8e195811debb821eb720ca16b798c477297a4bc
Author: ben fleis <benfleis@users.noreply.github.com>
Date:   Tue Jan 6 08:45:09 2026 +0100

    sync - most basic before exposing parse

commit e78837002ac4c74b75e16fef0a8f9eebb954a815
Merge: f24649ac8f 99a37abe11
Author: ben fleis <benfleis@users.noreply.github.com>
Date:   Mon Jan 5 14:04:53 2026 +0100

    Merge branch 'main' of github.com:duckdb/duckdb into sanitize-joinpath-tweaks

commit f24649ac8f32765750d4926784ec120a6408cb62
Merge: c61516aac0 c84d2d52f3
Author: ben fleis <benfleis@users.noreply.github.com>
Date:   Mon Dec 22 22:12:15 2025 +0100

    Merge branch 'sanitize-joinpath-tweaks' of github.com:benfleis/duckdb into sanitize-joinpath-tweaks

commit c61516aac01cc1c75abc0728b43f5bdc91fe83d2
Merge: 56e496ffc4 e5d4d5ae80
Author: ben fleis <benfleis@users.noreply.github.com>
Date:   Mon Dec 22 22:06:10 2025 +0100

    Merge branch 'sanitize-joinpath' of https://github.com/feichai0017/duckdb into sanitize-joinpath-tweaks

commit 56e496ffc40a3506edba73af466af926b03896a1
Author: ben fleis <benfleis@users.noreply.github.com>
Date:   Fri Dec 19 11:07:22 2025 +0100

    a few last tweaks from feedbacks on fs.JoinPath

    - fix "drive relative" `JP("c:", "foo") -> c:foo` (instead of `c:/foo`)
    - avoid strange edge case /foo/bar/sftp://baz ("legit" : with slashes)
    - resume follow classic NULL handling pattern
    - homogenize error text & complete tests on error conditions

commit 72b69ab6ad3a2c07df27d7299d8ed8c972f9a288
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Tue Nov 25 16:53:21 2025 +1100

    Sanitize JoinPath handling and add tests

    supports:
    - local files, *nix (/foo/bar) and win (c:\baz) style,
    - network URI style (https://fizz/buzz, az://here/there)
    - absolute & relative paths
    - extends functionality to new `path_join()` SQL function
    - fails on un-joinables (differs from some implementations which choose
      RHS when incompatible)

    Normalizes along the way, in slashes, single- and double-dots.

commit e5d4d5ae80351a3549acc3de0eb27be8378a9f5e
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Mon Dec 22 18:22:00 2025 +0800

    Fix Windows glob path joining for drive roots

commit d6901efa3ce2ef3e41c20e45ba2493a9b01d2ffc
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Mon Dec 22 15:59:55 2025 +0800

    Adjust path_join null handling and tests

commit 4559885b9d414ed18fe07b2b93891493aacc1ce6
Merge: e1deafe411 c46a01b579
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Mon Dec 22 13:13:34 2025 +0800

    Merge branch 'main' into sanitize-joinpath

commit c84d2d52f3fab8ca35b5da1f135a022bc0df788d
Author: ben fleis <benfleis@users.noreply.github.com>
Date:   Fri Dec 19 11:07:22 2025 +0100

    a few last tweaks from feedbacks on fs.JoinPath

    - fix "drive relative" `JP("c:", "foo") -> c:foo` (instead of `c:/foo`)
    - avoid strange edge case /foo/bar/sftp://baz ("legit" : with slashes)
    - resume follow classic NULL handling pattern
    - homogenize error text & complete tests on error conditions

commit a2ca0c0bf126f68e062db47cf939f3286918b340
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Tue Nov 25 16:53:21 2025 +1100

    Sanitize JoinPath handling and add tests

    supports:
    - local files, *nix (/foo/bar) and win (c:\baz) style,
    - network URI style (https://fizz/buzz, az://here/there)
    - absolute & relative paths
    - extends functionality to new `path_join()` SQL function
    - fails on un-joinables (differs from some implementations which choose
      RHS when incompatible)

    Normalizes along the way, in slashes, single- and double-dots.

commit e1deafe4114746a2aa851e2e817fcb92f3e6846c
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:58 2025 +0800

    update function generation

commit 3dd5db31724a333aef2aa40856b0b5427ccabd27
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:57 2025 +0800

    correct code format

commit 2650670108dac0cbbd4893cf06283001624a5cd8
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:56 2025 +0800

    Fix JoinPath root separator duplication

commit f5e4d8a5b43ababa62813d741751a0e89a50dd29
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:55 2025 +0800

    Format path_join description

commit cbae45db47c3150e9e06b4b17a5c9eeaf695c9a1
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:54 2025 +0800

    Clarify path_join strictness and Windows path handling

commit 13a46b70147a4286943433a9644c58f57d7518c8
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:53 2025 +0800

    Fix file scheme joins with absolute paths

commit 15ffeb5c9071b9d767586fa9af7b17c4a04a1ed3
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:52 2025 +0800

    Fix JoinPath URI separator handling

commit eba9f05a30dba34626e25c252088671c936917d8
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:51 2025 +0800

    Force forward slashes for URI joins regardless of platform separator

commit b52fad951a4be7cdaa3c12882ed3e5004865bb7e
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:50 2025 +0800

    Normalize URI paths to forward slashes when separator is slash

commit a698a8a77809a9ea0e066a8a4a4760b5839419e3
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:49 2025 +0800

    Normalize URI paths to forward slashes in JoinPath

commit fd000090d81356928b4224ee7a24bfb9ed3c5eb7
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:48 2025 +0800

    Revert "Keep URI paths using forward slashes when joining"

    This reverts commit 9b06a6250e3548d09abf4bb8bac8dcbe825d0434.

commit 3bdc7df1799c995454a999c0ddd5b200e72f38e7
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:16:46 2025 +0800

    Keep URI paths using forward slashes when joining

commit 7c121626ae66ea2116248e7df2a863fcc692cbe1
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:15:26 2025 +0800

    Relax drive-relative clamp and tidy separator stripping

commit a45efab3b507b59af626a408d9904083b2fddfd9
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Wed Dec 17 15:12:13 2025 +0800

    Fix JoinPath file scheme handling

commit 2296b6c771d04a47165191ca5bc895881607270b
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sun Nov 30 02:39:51 2025 +1100

    Stabilize path_join evaluation

commit 832ccb5be8fcdbde538526e87d8f15737f1efe38
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sat Nov 29 23:33:29 2025 +1100

    Make path_join SQL test platform-neutral

commit 1666b8100ba9206c2dcb9efb418455e8e73503c6
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sat Nov 29 20:45:25 2025 +1100

    Regenerate function metadata for path_join

commit 4394f8e63661d62652536211ae1a978f8bd39990
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sat Nov 29 20:38:04 2025 +1100

    fix code format

commit bb4944a033d8a02549fe5aef6f8922486f6e540b
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sat Nov 29 20:36:00 2025 +1100

    Add path_join SQL function

commit dae576807248984e5b25ce4474f66297fb7f4dcd
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sat Nov 29 10:29:37 2025 +1100

    correct coede format

commit 41175dcd3e6069b6304b67c438453a947058bdca
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Sat Nov 29 03:16:37 2025 +1100

    Fix Windows drive root joins and add glob regression test

commit 24cf2412f6f6daf1eae01684ea75c7935afd169c
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Fri Nov 28 22:42:49 2025 +1100

    Harden JoinPath normalization and edge tests

commit 909979054d0f14dc3a786cc4b2f5e85edd6e3989
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Tue Nov 25 17:09:50 2025 +1100

    fix code format

commit 2ae4636e734ad4be54b61833865daa830c321d58
Author: feichai0017 <songguocheng348@gmail.com>
Date:   Tue Nov 25 16:53:21 2025 +1100

    Sanitize JoinPath handling and add tests

benfleis committed:
many fixups, mostly tweaked semantics or existing bugs ...

... that were exposed by updated JoinPath call; also re-remove some tests
defunct tests that remained after the large rebase

- fix issues from recent MakePathAbsolute and glob changes
- introduce a couple convenience funcs to simplify the above, plus tests
- fix embedded separator assumption in log storage
- fix CreateDirectoriesRecursive (known issues w / and \)
- also fix other found separator issues, and add comment/test to Parent()
  to illustrate why it uses ".." instead of resize+math.
- explicit trailing separator support, which is meaningful in globs

benfleis reviewed (commented):
Self review - a few known tweaks, and callouts for @Mytherin to double check since most of the conflicts were related to his recent changes in path handling.

Mytherin commented on src/function/scalar/string/path_join.cpp:
```diff
@@ -0,0 +1,117 @@
+
+unique_ptr<FunctionData> PathJoinBind(ClientContext &context, ScalarFunction &bound_function,
+                                      vector<unique_ptr<Expression>> &arguments) {
+	if (arguments.empty()) {
```
Is this bind function required? We can determine the path count in the executor by just looking at the number of columns in the input chunk, no?

Mytherin commented on src/common/file_system.cpp:
```diff
@@ -280,9 +283,340 @@ string FileSystem::GetWorkingDirectory() {
+		vector<string> result;
+		result.reserve(lhs.segments.size());
+		for (auto &seg : lhs.segments) {
+			if (seg == ".." && !result.empty() && result.back() != "..") {
```
Should we normalize `./` as well? Or are we calling canonicalize after?

Mytherin commented on src/common/file_system.cpp:
```diff
@@ -280,9 +283,340 @@ string FileSystem::GetWorkingDirectory() {
+	return Join(dots);
+}
+
 string FileSystem::JoinPath(const string &a, const string &b) {
```
Not for this PR but I think it would be nice if this method was varargs as well

benfleis replied:
yes indeed, I have a list of follow ups, and that's in it, along with some other places in the code that would simplify by using this.

Mytherin reviewed (commented):
Thanks for the PR! Looks good - left some comments. I think we should target `main` with this as it's not really a bug-fix and has a substantial risk of causing issues.

benfleis commented on src/common/file_system.cpp:
```diff
@@ -280,9 +283,340 @@ string FileSystem::GetWorkingDirectory() {
 
+// Handle both forward & backward slashes in 1 func, hiding the _WIN32 #def
+static inline bool IsPathSeparator(char c) {
+	return (c == '/' || c == '\\');
```
Originally I had the 2nd == wrapped in _WIN32 as you might guess from the comment.

This code now crosses into SQL syntax, and it's unclear how often we might have users on non-windows duckdb instances parsing windows paths, this feels right.

My inclination is to retain the flexibility of both in all parsing contexts, and assume/expect that '\' in actual linux paths will not be a real issue.

(It _could_ be strictly handled at the LocalFileSystem layer with some API tweaks as well.)

Thoughts?

benfleis commented on src/common/file_system.cpp:
```diff
@@ -280,9 +283,340 @@ string FileSystem::GetWorkingDirectory() {
+
+// NOTE: instead of chopping off segments, apply '..'; this guards against traps with odd paths (leading ..,
+// etc.) to make sure we always handle Parentage consistently, at a slight cost of cpu/tmp allocation.
+Path Path::Parent(int n) const {
```
yep, I'd love to.

benfleis commented on src/common/file_system.cpp:
```diff
@@ -280,9 +283,340 @@ string FileSystem::GetWorkingDirectory() {
+}
+
+bool Path::HasDrive() const {
+	return GetDriveChar() != 0;
```
I think the function should stand -- any non-windows path HasDrive -> false seems fine and avoids having a fundamentally different API per arch.

benfleis committed:
feedback + follow ups

- extract Path -> path.[ch]pp, privatize members and add a handful
  Is/Has/Get accessors
- add variadic & vector Join funcs (match SQL)
- cleanup/add to the SQL .test path_join tests (correct '' defaulting)

benfleis committed:
drop unnecessary Bind function

benfleis committed:
use args.AllConstant instead of computing self

benfleis committed:
small fixes + linux compiler complaint
